### PR TITLE
Implement 'lxc image edit'

### PR DESCRIPTION
**Pull request by hallyn**
Signed-off-by: Serge Hallyn <serge.hallyn@ubuntu.com>

### CODE REVIEW

File: lxc/config.go [hunk 1]
```diff
@@ -308,6 +333,51 @@ func doProfileCreate(client *lxd.Client, p string) error {
...
+	profile, err := client.ProfileConfig(p)
+	editor := os.Getenv("VISUAL")
+	if editor == "" {
+		editor = os.Getenv("EDITOR")
+		if editor == "" {
+			editor = "/usr/bin/vi"
+		}
+	}
+	data, err := yaml.Marshal(&profile)
+	f, err := ioutil.TempFile("/tmp", "image_")
```
Comment (stgraber) [inline lxc/config.go]: Should we perhaps include "lxc" in the tempfile prefix here?

File: lxc/image.go [hunk 2]
```diff
@@ -184,6 +206,62 @@ func (c *imageCmd) run(config *lxd.Config, args []string) error {
...
+		d, err := lxd.NewClient(config, remote)
+		if err != nil {
+			return err
+		}
+		properties, err := d.GetImageProperties(image)
+		editor := os.Getenv("VISUAL")
+		if editor == "" {
+			editor = os.Getenv("EDITOR")
+			if editor == "" {
+				editor = "/usr/bin/vi"
```
Comment (stgraber) [inline lxc/image.go]: Doesn't command support path lookups?

File: lxc/image.go [hunk 3]
```diff
@@ -184,6 +206,62 @@ func (c *imageCmd) run(config *lxd.Config, args []string) error {
...
+			}
+		}
+		data, err := yaml.Marshal(&properties)
+		f, err := ioutil.TempFile("/tmp", "image_")
```
Comment (stgraber) [inline lxc/image.go]: Same comment as above.

File: lxc/config.go [hunk 4]
```diff
@@ -308,6 +333,51 @@ func doProfileCreate(client *lxd.Client, p string) error {
 	return err
 }
 
+func doProfileEdit(client *lxd.Client, p string) error {
+	profile, err := client.ProfileConfig(p)
+	editor := os.Getenv("VISUAL")
+	if editor == "" {
+		editor = os.Getenv("EDITOR")
+		if editor == "" {
+			editor = "/usr/bin/vi"
```
Comment (stgraber) [inline lxc/config.go]: Same comment as below. (path lookup)

File: lxd/images.go [hunk 5]
```diff
@@ -304,21 +304,95 @@ var imagesCmd = Command{name: "images", post: imagesPost, get: imagesGet}
...
-		shared.Debugf("Error deleting image file %s: %s\n", uuidfname, err)
+		shared.Debugf("Error deleting image file %s: %s\n", fname, err)
 	}
 
-	_, _ = d.db.Exec("DELETE FROM images_aliases WHERE image_id=(SELECT id FROM images WHERE fingerprint=?);", uuid)
-	_, _ = d.db.Exec("DELETE FROM images_properties WHERE image_id=(SELECT id FROM images WHERE fingerprint=?);", uuid)
-	_, _ = d.db.Exec("DELETE FROM images WHERE fingerprint=?", uuid)
+	_, _ = d.db.Exec("DELETE FROM images_aliases WHERE image_id=(SELECT id FROM images WHERE fingerprint=?);", fingerprint)
+	_, _ = d.db.Exec("DELETE FROM images_properties WHERE image_id=(SELECT id FROM images WHERE fingerprint=?);", fingerprint)
+	_, _ = d.db.Exec("DELETE FROM images WHERE fingerprint=?", fingerprint)
```
Comment (stgraber) [inline lxd/images.go]: I know it's not getting us much, but would you mind doing the "SELECT id from IMAGES where fingerprint=?" separately and then use the value in the two DELETEs? That'd save us one SQL query.

**Comment by stgraber** [comment]
Also, do you have a reason for sometimes picking single quotes vs regular quotation marks?
I typically use regular quotation marks for strings passed to a function and single quotes for a dictionary index but I don't know if we're following some other styling rule in LXD.

**Comment by hallyn** [comment]
On Mon, Mar 09, 2015 at 08:50:31AM -0700, St�phane Graber wrote:
> Also, do you have a reason for sometimes picking single quotes vs regular quotation marks?
> I typically use regular quotation marks for strings passed to a function and single quotes for a dictionary index but I don't know if we're following some other styling rule in LXD.

I use ` for lines which I expect to become multi-line as I start typing them.

File: lxc/config.go (hunk 1 above)
Comment (tych0) [inline lxc/config.go]: On Mon, Mar 09, 2015 at 08:44:17AM -0700, St�phane Graber wrote:
> > @@ -308,6 +333,51 @@ func doProfileCreate(client *lxd.Client, p string) error {
> >  	return err
> >  }
> >  
> > +func doProfileEdit(client *lxd.Client, p string) error {
> > +	profile, err := client.ProfileConfig(p)
> > +	editor := os.Getenv("VISUAL")
> > +	if editor == "" {
> > +		editor = os.Getenv("EDITOR")
> > +		if editor == "" {
> > +			editor = "/usr/bin/vi"
> > +		}
> > +	}
> > +	data, err := yaml.Marshal(&profile)
> > +	f, err := ioutil.TempFile("/tmp", "image_")
> 
> Should we perhaps include "lxc" in the tempfile prefix here?

Not sure, but we should probably (?) remove "/tmp" and just make it
""; golang is smart enough to respect environment variables and do the
right thing on different platforms, so if we ever want to put
temporary files somewhere else, we'd have to change the code.

**Comment by stgraber** [comment]
Oh, sorry, I thought they were single quotes, not backtick. Backtick for multi-line does make sense indeed.

File: lxc/config.go [hunk 6]
```diff
@@ -17,6 +21,24 @@ func (c *configCmd) showByDefault() bool {
...
+### using the 'edit' command.  See 'lxc profile move' for renaming.
+###
+### After the name comes the 'config:' section header, followed by configuration
+### items, one per line.
+###
+### Next come the devices.  Since each device has optional properties, the
+### devices are each one sub-section followed by properties listed on separate
+### lines as 'key: value'
+
+`
```
Comment (stgraber) [inline lxc/config.go]: I suspect we want that translated, so we shouldn't use backtick and wrap this with Gettext.

File: lxc/image.go [hunk 7]
```diff
@@ -14,21 +19,38 @@ func (c *imageCmd) showByDefault() bool {
...
+###  The first is 'imagetype: ' followed by an integer.  0 means
+###  a short string, 1 means a long text value containing newlines.
+###
+###  The second is the key, which is the property name, for instance
+###  name, os, or description.
+###
+###  The third is the value, for instance 'ubuntu-14.04-amd64-20150218',
+###  'Ubuntu', or 'Ubuntu 14.04 LTS Intel 64bit'
+
+`
```
Comment (stgraber) [inline lxc/image.go]: I suspect we want that translated, so we shouldn't use backtick and wrap this with Gettext.

**Pull request merged**